.github/workflows/hw_mode_test.yml

```diff
@@ -948,3 +948,63 @@ jobs:
       run: |
         docker exec ${{ env.CONTAINER_NAME }} bash -c "cat /root/occlum/build/test/.fail"
         docker stop ${{ env.CONTAINER_NAME }}
+  
+  Spark_test:
+    if: github.event_name == 'push' || ${{ contains(github.event.pull_request.labels.*.name, 'SGX-hardware-test-required') }}
+    runs-on: ${{ matrix.self_runner }}
+    strategy:
+      matrix:
+        self_runner: [[self-hosted, SGX2-HW]]
+
+    steps:
+    - name: Clean before running
+      run: |
+        sudo chown -R ${{ secrets.CI_ADMIN }} "${{ github.workspace }}"
+
+    - name: Checkout code
+      if: github.event_name == 'push'
+      uses: actions/checkout@v2
+      with:
+        submodules: true
+
+    - name: Checkout code from fork
+      if: ${{ contains(github.event.pull_request.labels.*.name, 'SGX-hardware-test-required') }}
+      uses: actions/checkout@v2
+      with:
+        ref: refs/pull/${{ github.event.pull_request.number }}/merge
+        submodules: true
+    
+    - name: Create container
+      run: |
+        docker pull occlum/occlum:latest-ubuntu20.04
+        if [[ "${{ matrix.self_runner[1] }}" == "SGX1-HW" ]]; then
+          spark_test=$(docker run -itd --privileged --rm --env CARGO_HTTP_MULTIPLEXING=false --device /dev/isgx -v $GITHUB_WORKSPACE:/root/occlum -v $(which docker):/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock occlum/occlum:latest-ubuntu20.04);
+        elif [[ "${{ matrix.self_runner[1] }}" == "SGX2-HW" ]]; then
+          spark_test=$(docker run -itd --privileged --rm --env CARGO_HTTP_MULTIPLEXING=false -v /dev/sgx:/dev/sgx -v $GITHUB_WORKSPACE:/root/occlum -v $(which docker):/usr/bin/docker -v /var/run/docker.sock:/var/run/docker.sock occlum/occlum:latest-ubuntu20.04);
+        else
+          echo "Unsupported Hardware"
+        fi;
+        echo "spark_test=$spark_test" >> $GITHUB_ENV
+
+    - name: Change download source of crates.io
+      run: |
+        docker exec $spark_test bash -c "cat <<- EOF >/root/.cargo/config
+        [source.crates-io]
+        registry = \"https://github.com/rust-lang/crates.io-index\"
+        replace-with = 'ustc'
+        [source.ustc]
+        registry = \"git://mirrors.ustc.edu.cn/crates.io-index\"
+        EOF"
+
+    - name: Build dependencies
+      run: docker exec $spark_test bash -c "cd /root/occlum; make submodule"
+
+    - name: Make install
+      run: docker exec $spark_test bash -c "source /opt/intel/sgxsdk/environment; cd /root/occlum; OCCLUM_RELEASE_BUILD=1 make install"
+
+    - name: Prepare env
+      run: docker exec $spark_test bash -c "cd /root/occlum/demos/spark && ./prepare_env.sh"
+    
+    - name: Run spark pi
+      run: docker exec $spark_test bash -c "cd /root/occlum/demos/spark && ./run_spark.sh pi"
+      
```

Comment:
Check the file ending here. Looks like something is wrong.

Reply:
Done